src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1019,11 +1020,70 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
     ucs_mpool_put(skb);
 }
 
-ucs_status_t uct_ud_ep_flush_nolock(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
-                                    uct_completion_t *comp)
+static ucs_status_t
+uct_ud_ep_comp_skb_add(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
+                       uct_completion_t *comp)
+{
+    uct_ud_send_skb_t *skb;
+
+    if (comp == NULL) {
+        return UCS_INPROGRESS;
+    }
+
+    /* If the user requested a callback, allocate a dummy skb which will be
+     * released when the current sequence number is completed.
+     */
+    ucs_assert(comp->count > 0);
+
+    skb = ucs_mpool_get(&iface->tx.mp);
+    if (skb == NULL) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    /* Add dummy skb to the window, which would call user completion
+     * callback when getting ACK.
+     */
+    skb->flags                  = UCT_UD_SEND_SKB_FLAG_COMP;
+    skb->len                    = sizeof(skb->neth[0]);
+    skb->neth->packet_type      = 0;
+    skb->neth->psn              = (uct_ud_psn_t)(ep->tx.psn - 1);
+    uct_ud_neth_set_dest_id(skb->neth, UCT_UD_EP_NULL_ID);
+    uct_ud_comp_desc(skb)->comp = comp;
+
+    if (!ucs_queue_is_empty(&ep->tx.window)) {
+        /* If window non-empty: add to window */
+        skb->flags |= UCT_UD_SEND_SKB_FLAG_ACK_REQ;
+        ucs_queue_push(&ep->tx.window, &skb->queue);
+    } else {
+        /* Otherwise, add the skb after async completions */
+        ucs_assert(ep->tx.resend_count == 0);
+        uct_ud_iface_add_async_comp(iface, ep, skb, UCS_OK);
+    }
+
+    ucs_trace_data("added dummy flush skb %p psn %d user_comp %p", skb,
+                   skb->neth->psn, comp);
+
+    return UCS_INPROGRESS;
+}
+
+static UCS_F_ALWAYS_INLINE int
+uct_ud_ep_should_send_ack_req(const uct_ud_ep_t *ep, unsigned flags)
 {
     uct_ud_send_skb_t *skb;
 
+    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
```

Comment:
1. if we are doing flush cancel and window is non empty and last skb has flag UCT_UD_SEND_SKB_FLAG_ACK_REQ, do we still need to send another skb with the flag UCT_UD_SEND_SKB_FLAG_ACK_REQ?
2. don't we need to mark last skb in the window with UCT_UD_SEND_SKB_FLAG_ACK_REQ, even if comp==NULL?

Reply:
> if we are doing flush cancel and window is non empty and last skb has flag UCT_UD_SEND_SKB_FLAG_ACK_REQ, do we still need to send another skb with the flag UCT_UD_SEND_SKB_FLAG_ACK_REQ?

yes, we should send it.
e.g. if peer was keelied and we did flush(LOCAL) and CQE was already read, but not ACKed by peer - SKB won't completed. and if the next flush(CANCEL) won't post ACK_REQ which will generate CQE - neither flush(LOCAL) nor flush(CANCEL) won't be completed until resend attemts aren't completed.

> don't we need to mark last skb in the window with UCT_UD_SEND_SKB_FLAG_ACK_REQ, even if comp==NULL?

done